### PR TITLE
Paginate column access in AstyanaxStore

### PR DESCRIPTION
**Pull request by dpitera**
#411 

JanusGraph has a tendency to make queries to Cassandra with _very_ large
limits. For example, any query without a `.limit()` clause will fetch
all matching columns from a given row. These requests are usually
range-bound (e.g., only requesting columns that are edges, based on the
serialization logic) but are not limited.

Fetching millions of columns in a single request is not very friendly to
Cassandra; in production we've observed memory allocation spikes in
excess of 10GB in a 30s period. On my local machine, calling

graph.traversal().V(id).outE().count()

prior to this commit will cause my Cassandra instance to OOM in
approximately 10 seconds.

This patch places a hard cap on the number of columns returned (per row)
via a single request to Cassandra. If, for a given request, a row has
more columns available in the requested range than the cap, the system
makes additional requests for that row until the requested number of
columns is returned or no more columns are available.

This code is somewhat more subtle that may be expected due to Titan's
use of the Astyanax API - it's not uncommon for Titan to make a request
for multiple keys concurrently, in which case the `limit` parameter
applies to each key - not the request as a whole. This has two
implications:

1.  We need to use a different API (getRow, rather than getKeySlice) to
fetch columns for rows that met the limit on the initial query, since
the starting column must be manipulated on a per-row basis

2. The page size needs to be relatively conservative, in case a single
request is made for hundreds of keys at once. This is relatively
unlikely, especially given the natural infrequency of supernodes,
but it's something that we must be aware of.

Signed-off-by: David Pitera <dpitera@us.ibm.com>

### CODE REVIEW

**Comment by dpitera** [comment]
Pardon my naiveté, but what sort of test are you looking for other than the passing of the cassandra astyanax test suite?

**Comment by sjudeng** [comment]
I'd expect to see tests with smaller page size or larger test graphs (in terms of properties+edges since it's column pagination) to make sure the pagination implementation is being exercised. I'd also look at testing boundary conditions (e.g. number of columns equals page size).

**Comment by dpitera** [comment]
Excellent suggestions, tests have been added!

File: janusgraph-cassandra/src/main/java/org/janusgraph/diskstorage/cassandra/astyanax/AstyanaxKeyColumnValueStore.java [hunk 1]
```diff
@@ -133,8 +146,69 @@ public EntryList getSlice(KeySliceQuery query, StoreTransaction txh) throws Back
 
         for (Row<ByteBuffer, ByteBuffer> row : rows) {
             assert !result.containsKey(row.getKey());
-            result.put(StaticArrayBuffer.of(row.getKey()),
-                  CassandraHelper.makeEntryList(row.getColumns(),entryGetter, query.getSliceEnd(), query.getLimit()));
+            ByteBuffer key = row.getKey();
+            ColumnList<ByteBuffer> pageColumns = row.getColumns();
+            ArrayList<Column<ByteBuffer>> queryColumns = new ArrayList();
```
Comment (sjudeng) [inline janusgraph-cassandra/src/main/java/org/janusgraph/diskstorage/cassandra/astyanax/AstyanaxKeyColumnValueStore.java]: `List` on LHS

File: janusgraph-cassandra/src/main/java/org/janusgraph/diskstorage/cassandra/astyanax/AstyanaxKeyColumnValueStore.java [hunk 2]
```diff
@@ -110,16 +113,26 @@ public EntryList getSlice(KeySliceQuery query, StoreTransaction txh) throws Back
...
-                .withRetryPolicy(retryPolicy.duplicate())
-                .getKeySlice(CassandraHelper.convert(keys));
 
-        // Thank you, Astyanax, for making builder pattern useful :(
-        rq.withColumnRange(query.getSliceStart().asByteBuffer(),
-                query.getSliceEnd().asByteBuffer(),
-                false,
-                query.getLimit() + (query.hasLimit()?1:0)); //Add one for potentially removed last column
+         // Add one for last column potentially removed in CassandraHelper.makeEntryList
+        int queryLimit = query.getLimit() + (query.hasLimit() ? 1 : 0);
```
Comment (sjudeng) [inline janusgraph-cassandra/src/main/java/org/janusgraph/diskstorage/cassandra/astyanax/AstyanaxKeyColumnValueStore.java]: `final` here and throughout wherever possible

File: janusgraph-cassandra/src/main/java/org/janusgraph/diskstorage/cassandra/astyanax/AstyanaxKeyColumnValueStore.java [hunk 3]
```diff
@@ -133,8 +146,69 @@ public EntryList getSlice(KeySliceQuery query, StoreTransaction txh) throws Back
...
+            Iterables.addAll(queryColumns, pageColumns);
+            while (pageColumns.size() == pageLimit && queryColumns.size() < queryLimit) {
+                Column<ByteBuffer> lastColumn = queryColumns.get(queryColumns.size() - 1);
+                sliceStart = lastColumn.getName();
+                // No possibility of two values at the same column name, so start the
+                // next slice one bit after the last column found by the previous query.
+                // byte[] is little-endian
+                Integer position = null;
+                for (int i = sliceStart.array().length - 1; i >= 0; i--) {
+                    if (sliceStart.array()[i] < 127) {
```
Comment (sjudeng) [inline janusgraph-cassandra/src/main/java/org/janusgraph/diskstorage/cassandra/astyanax/AstyanaxKeyColumnValueStore.java]: /127/Byte.MAX_VALUE/

File: janusgraph-cassandra/src/main/java/org/janusgraph/diskstorage/cassandra/astyanax/AstyanaxKeyColumnValueStore.java [hunk 4]
```diff
@@ -133,8 +146,69 @@ public EntryList getSlice(KeySliceQuery query, StoreTransaction txh) throws Back
...
+
+                OperationResult<ColumnList<ByteBuffer>> pageResult;
+                try {
+                    pageResult = (OperationResult<ColumnList<ByteBuffer>>) pageQuery.execute();
+                } catch (ConnectionException e) {
+                    throw new TemporaryBackendException(e);
+                }
+
+                if (Thread.interrupted()) {
+                    throw new RuntimeException(new InterruptedException());
```
Comment (sjudeng) [inline janusgraph-cassandra/src/main/java/org/janusgraph/diskstorage/cassandra/astyanax/AstyanaxKeyColumnValueStore.java]: Should this be a `TraversalInterruptedException`?

File: janusgraph-cassandra/src/main/java/org/janusgraph/diskstorage/cassandra/astyanax/AstyanaxStoreManager.java [hunk 5]
```diff
@@ -270,6 +279,7 @@
 
     private final RetryPolicy retryPolicy;
 
+    final int readPageSize;
```
Comment (amcp) [inline janusgraph-cassandra/src/main/java/org/janusgraph/diskstorage/cassandra/astyanax/AstyanaxStoreManager.java]: private

File: janusgraph-cassandra/src/test/java/org/janusgraph/diskstorage/cassandra/astyanax/AstyanaxColumnPaginationTest.java [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+    @Test
+    public void ensureReadPageSizePropertySetCorrectly() {
+        assertEquals(((AstyanaxStoreManager) manager).readPageSize, 4096);
+    }
+
+    @Test
+    public void retrieveLessThanBoundaryColumnPaginationProperties() {
+        Graph graph = JanusGraphFactory.open(getBaseStorageConfiguration());
+        Vertex v = graph.addVertex();
+        for(int i = 0; i < 4095; i++) {
```
Comment (amcp) [inline janusgraph-cassandra/src/test/java/org/janusgraph/diskstorage/cassandra/astyanax/AstyanaxColumnPaginationTest.java]: refactor 4095 and reuse here and below

File: janusgraph-cassandra/src/test/java/org/janusgraph/diskstorage/cassandra/astyanax/AstyanaxColumnPaginationTest.java [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+        graph.tx().commit();
+
+        assertEquals(4097, graph.traversal().V(v).valueMap().next().keySet().size());
+    }
+
+    @Test
+    public void retrieveWayBeyondBoundaryColumnPaginationProperties() {
+        Graph graph = JanusGraphFactory.open(getBaseStorageConfiguration());
+        Vertex v = graph.addVertex();
+        for (int i = 0; i < 20000; i++) {
```
Comment (amcp) [inline janusgraph-cassandra/src/test/java/org/janusgraph/diskstorage/cassandra/astyanax/AstyanaxColumnPaginationTest.java]: refactor a constant for 20000

File: janusgraph-cassandra/src/main/java/org/janusgraph/diskstorage/cassandra/astyanax/AstyanaxStoreManager.java (hunk 5 above)
Comment (dpitera) [inline janusgraph-cassandra/src/main/java/org/janusgraph/diskstorage/cassandra/astyanax/AstyanaxStoreManager.java]: this has no modifier on purpose to allow access for the `AstyanaxKeyColumnValueStore` in the same package.

**Comment by dpitera** [comment]
@sjudeng @amcp Updated with your feedback. Thanks for your reviews

**Comment by sjudeng** [comment]
I'm not seeing the updates. Did you commit/push?

**Comment by dpitera** [comment]
Sorry I don't know what happened; all updates should be in place now.
